daemons/ipa-kdb/ipa_kdb_kdcpolicy.c

```diff
@@ -0,0 +1,159 @@
+                       const krb5_db_entry *client,
+                       const krb5_db_entry *server,
+                       const char *const *auth_indicators,
+                       const char **status, krb5_deltat *lifetime_out,
+                       krb5_deltat *renew_lifetime_out)
+{
+    krb5_error_code kerr;
+    enum ipadb_user_auth ua;
+    struct ipadb_e_data *ied;
+    int valid_auth_indicators = 0;
+
+    *status = NULL;
+    *lifetime_out = 0;
+    *renew_lifetime_out = 0;
+
+    krb5_klog_syslog(LOG_INFO, "IPA kdcpolicy: checking AS-REQ.");
+
+    ied = (struct ipadb_e_data *)client->e_data;
+    if (ied == NULL || ied->magic != IPA_E_DATA_MAGIC) {
+        /* e-data is not availble, getting user auth from LDAP */
+        krb5_klog_syslog(LOG_INFO, "IPA kdcpolicy: client e_data not availble. Try fetching...");
+        kerr = ipadb_get_principal(context, request->client, KRB5_KDB_FLAG_ALIAS_OK, &client);
+        if (kerr != 0) {
+            krb5_klog_syslog(LOG_ERR, "IPA kdcpolicy: ipadb_find_principal failed.");
+            return kerr;
+        }
+
+        ied = (struct ipadb_e_data *)client->e_data;
+        if (ied == NULL && ied->magic != IPA_E_DATA_MAGIC) {
+            krb5_klog_syslog(LOG_ERR, "IPA kdcpolicy: client e_data fetching failed.");
+            return EINVAL;
+        }
+    }
+
+    ua = ied->user_auth;
+    
+    /* If no mechanisms are set, allow every auth method */
+    if (ua == IPADB_USER_AUTH_NONE) {
+        return 0;
+    }
+
+    /* For each auth indicator, see if it is allowed for that user */
+    for (int i = 0; auth_indicators[i] != NULL; i++) {
+        const char *auth_indicator = auth_indicators[i];
+
+        if (strcmp(*auth_indicators, "otp") == 0) {
+            valid_auth_indicators++;
+            if (!(ua & IPADB_USER_AUTH_OTP)) {
+                *status = "OTP pre-authentication not allowed for this user.";
+                return KRB5KDC_ERR_POLICY;
+            }
+        } else if (strcmp(*auth_indicators, "radius") == 0) {
+            valid_auth_indicators++;
+            if (!(ua & IPADB_USER_AUTH_RADIUS)) {
+                *status = "OTP pre-authentication not allowed for this user.";
+                return KRB5KDC_ERR_POLICY;
+            }
+        } else if (strcmp(*auth_indicators, "pkinit") == 0) {
+            valid_auth_indicators++;
+            if (!(ua & IPADB_USER_AUTH_PKINIT)) {
+                *status = "PKINIT pre-authentication not allowed for this user.";
+                return KRB5KDC_ERR_POLICY;
+            }
+        } else if (strcmp(*auth_indicators, "hardened") == 0) {
+            valid_auth_indicators++;
+            /* Allow hardened even if only password pre-auth is allowed */
+            if (!(ua & (IPADB_USER_AUTH_HARDENED | IPADB_USER_AUTH_PASSWORD))) {
+                *status = "Password pre-authentication not not allowed for this user.";
+                return KRB5KDC_ERR_POLICY;
+            }
+        }
```

Comment:
So I assume we do not have an indicator for regular password based pre-auth, so I would add a comment to that effect in the code, and some handling if ua == IPADB_USER_AUTH_PASSWORD for some reason, otherwise this code will prevent a user with just the password ua and not indicators from being able to proceed at all.

Reply:
Ah I see that IPADB_USER_AUTH_PASSWORD is handled below as a special case ... uhmmm storngly advise some comment here that explains the reason, otherwise it will be hard to figure it out down the road once we forgot why.

Reply:
Comment added.